### DISCUSSION TITLE
How to get the usb device associated with a 'remove' event?

### DISCUSSION THREAD

**Discussion by brianreinhold** [pyudev/pyudev#476] [Q&A]
On a 'bind' event for a USB device I can get the VID, PID, and serial number from
```
        vendor_id: str = device.get('ID_VENDOR_ID')
        device_id: str = device.get('ID_MODEL_ID')
        serial_no: str = device.get('ID_SERIAL')
```
However, on an unbind or remove event all these fields give None. Since I may have multiple USB devices connected simultaneously, how can I identify the USB device that was unplugged and responsible for that remove event?

**Comment by mulkieran** [comment]
pyudev supports a monitor facility, which would allow you to catch the udev remove or unbind event resulting from the device's removal.

Otherwise, you would have to keep a cache of devices, obtained via ```list_devices```, and re-run ```list_devices``` at intervals to update your cache.

Or, you can rely on a property like ```device_node```; assuming that if your USB device's ```device_node()``` is None it has been removed seems reasonable.

**Comment by brianreinhold** [comment]
It turns out that there is a bug in the Raspberry PI kernel I am using where the PID, VID, and serial number are not properly propagated on a remove event. It is not a problem in pyudev. However, in lieu of that, I am using the device_path as a unique identifier of my map of device handlers, and i remove that entry as soon as I get the remove event. I have been looking at the four fields

device_path, system_path, device_number, and device_node. It looks like I could use any of these fields to correlate a plugin to an unplug event (well I am not sure about the device_number as I am not really sure what that represents).

At this stage it is working using the device_path.
